### PR TITLE
Add module definition; standardize method invocation

### PR DESCRIPTION
**Pull request by dbennett-sentry**

### CODE REVIEW

File: devenv/bootstrap.py [hunk 1]
```diff
@@ -13,22 +12,32 @@
...
+from devenv.lib.modules import ExitCode
+
+
+def main(context: Context, argv: Sequence[str] | None = None) -> ExitCode:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-d",
+        "--default-config",
+        action="append",
+        help="Provide a default config value. e.g., -s coderoot:path/to/root",
```
Comment (asottile-sentry) [inline devenv/bootstrap.py:27]: `-s` vs. `-d`

File: devenv/bootstrap.py [hunk 2]
```diff
@@ -13,22 +12,32 @@
...
+    )
 
-help = "Bootstraps the development environment."
-ExitCode: TypeAlias = "str | int | None"
+    args = parser.parse_args(argv)
 
+    configs = {
+        k: v for k, v in [i.split(":") for i in args.default_config or []]
```
Comment (asottile-sentry) [inline devenv/bootstrap.py:33]: probably `.split(1)` so that values can contain `:`

File: devenv/lib/modules.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    help: str
+    action: Action
+
+
+def require(var: str, message: str) -> Callable[[Action], Action]:
+    def outer(main: Action) -> Action:
+        def inner(context: Context, args: Sequence[str] | None) -> ExitCode:
+            if context.get(var) is None:
+                print(message)
+                return 1
```
Comment (asottile-sentry) [inline devenv/lib/modules.py:28]: this can be `raise SystemExit(message)` since that'll handle unwinding and also conveniently prints the message!

File: devenv/main.py [hunk 4]
```diff
@@ -67,47 +33,48 @@ def devenv(argv: Sequence[str], config_path: str) -> ExitCode:
...
+    for name, info in modinfo_list:
+        # Argparse stuff
+        child = subparser.add_parser(info.command, help=info.help)
+        child.add_argument(
+            "--xixax",
+            dest="_module_exec",
+            default=info.action,
+            required=False,
+            help=argparse.SUPPRESS,
         )
```
Comment (asottile-sentry) [inline devenv/main.py:62]: this is clever -- but very hard to follow -- here's a simpler alternative that doesn't require a secret argument:

```python
modules = {
    mod.name: mod for mod in [...]
}
return modules[args.command].whatever(...)
```

File: devenv/main.py [hunk 5]
```diff
@@ -67,47 +33,48 @@ def devenv(argv: Sequence[str], config_path: str) -> ExitCode:
...
+        "repo": Repository(current_root) if current_root else None,
+    }
 
-    if args.command == "doctor":
-        return doctor.main(context, remainder)
-    if args.command == "sync":
-        return sync.main(context, remainder)
-
-    return 1
+    return cast(ExitCode, args._module_exec(context, remainder))
```
Comment (asottile-sentry) [inline devenv/main.py:73]: `cast` has runtime overhead and generally should only be used for bugs in the type checker -- https://www.notion.so/sentry/sentry-typing-guidelines-best-practices-471d5aacf262476188982ce70f909e00?pvs=4#4c5f66d5e5f2460991cf1074111fdd85

File: devenv/bootstrap.py (hunk 1 above)
Comment (dbennett-sentry) [reply devenv/bootstrap.py:27]: Ack!

File: devenv/main.py (hunk 5 above)
Comment (dbennett-sentry) [reply devenv/main.py:73]: I think I picked this up from sync(). I'll remove it and check to make sure it still works.

File: devenv/main.py [hunk 6]
```diff
@@ -67,47 +33,49 @@ def devenv(argv: Sequence[str], config_path: str) -> ExitCode:
...
-    if args.command == "doctor":
-        return doctor.main(context, remainder)
-    if args.command == "sync":
-        return sync.main(context, remainder)
-
-    return 1
+    command_actions = {info.command: info.action for info in modinfo_list}
+    action = command_actions.get(args.command)
+    assert action is not None
+    return action(context, remainder)
```
Comment (IanWoodard) [inline devenv/main.py:74]: Played around with this a bit and found it worked quite well. One area that could use improvement is the help messages which give very little if any useful information (for specific actions). This doesn't have to be addressed in this PR, but something to keep an eye on.

**Review by IanWoodard — APPROVED** [review]

File: devenv/doctor.py [hunk 7]
```diff
@@ -173,14 +175,11 @@ def main(context: Dict[str, str], argv: Sequence[str] | None = None) -> int:
     )
     args = parser.parse_args(argv)
 
-    match_tags: set[str] = set(args.tag if args.tag else ())
-
     repo = context["repo"]
-    if repo not in {"sentry", "getsentry", "devenv"}:
-        print(f"repo {repo} not supported yet!")
-        return 1
+    assert repo is not None
```
[the remark this answers is not in the document]
Comment (dbennett-sentry) [reply devenv/doctor.py:179]: This is to inform the type checker that it is not None. If it were None, we'd have problems anyway so I'm using an assert instead of an if.
